internal/provider/network/firewall_resource.go

```diff
@@ -0,0 +1,504 @@
+		return
+	}
+
+	plan, d = r.read(ctx)
+	resp.Diagnostics.Append(d...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Set state to fully populated data
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+}
+
+// Delete deletes the resource and removes the Terraform state on success.
+func (r *firewallResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	state := &firewallModel{}
+
+	// Get current state
+	resp.Diagnostics.Append(req.State.Get(ctx, state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Init the resource
+	resp.Diagnostics.Append(r.Init(ctx, state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	/*
+		Implement the resource deletion here
+	*/
+
+	vdcOrVDCGroup, err := r.edgegw.GetParent()
+	if err != nil {
+		resp.Diagnostics.AddError("Error retrieving Edge Gateway parent", err.Error())
+		return
+	}
+
+	networkMutexKV.KvLock(ctx, vdcOrVDCGroup.GetID())
+	defer networkMutexKV.KvUnlock(ctx, vdcOrVDCGroup.GetID())
+
+	fwRules, err := r.edgegw.GetNsxtFirewall()
+	if err != nil {
+		resp.Diagnostics.AddError("Error retrieving Edge Gateway Firewall", err.Error())
+		return
+	}
+
+	if err := fwRules.DeleteAllRules(); err != nil {
+		resp.Diagnostics.AddError("Error deleting Edge Gateway Firewall", err.Error())
+		return
+	}
+}
+
+func (r *firewallResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	var (
+		edgegwID   string
+		edgegwName string
+		d          diag.Diagnostics
+	)
+
+	r.org, d = org.Init(r.client)
+	if d.HasError() {
+		resp.Diagnostics.Append(d...)
+		return
+	}
+
+	if uuid.IsValid(req.ID) {
+		edgegwID = uuid.Normalize(uuid.Gateway, req.ID).String()
+	} else {
+		edgegwName = req.ID
+	}
+
+	edgegw, err := r.org.GetEdgeGateway(edgegw.BaseEdgeGW{
+		ID:   types.StringValue(edgegwID),
+		Name: types.StringValue(edgegwName),
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to import firewall.", err.Error())
+		return
+	}
+
+	// ID
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("id"), edgegw.GetID())...)
+	// edge_gateway_id
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("edge_gateway_id"), edgegw.GetID())...)
+	// edge_gateway_name
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("edge_gateway_name"), edgegw.GetName())...)
+}
+
+func (r *firewallResource) read(ctx context.Context) (plan *firewallModel, diags diag.Diagnostics) {
+	fwRules, err := r.edgegw.GetNsxtFirewall()
+	if err != nil {
+		diags.AddError("Error retrieving Edge Gateway Firewall", err.Error())
+		return
+	}
+
+	rules, d := fwRulesRead(ctx, fwRules)
+	diags.Append(d...)
+	if diags.HasError() {
+		return
+	}
+
+	plan = &firewallModel{
+		ID:              types.StringValue(r.edgegw.GetID()),
+		EdgeGatewayID:   types.StringValue(r.edgegw.GetID()),
+		EdgeGatewayName: types.StringValue(r.edgegw.GetName()),
```

Comment:
```suggestion
                // ID is stored as Edge Gateway ID - because this is a "container" for all firewall rules at once and each child
               // TypeSet element will have a computed ID field for each rule
		ID:              types.StringValue(r.edgegw.GetID()),
		EdgeGatewayID:   types.StringValue(r.edgegw.GetID()),
		EdgeGatewayName: types.StringValue(r.edgegw.GetName()),
```